fix: make dependency cache key deterministic

jparise opened the pull request:
ReportTask could produce inconsistent cache keys for the same set of
dependencies because the key was neither deduplicated nor sorted. This
could cause both unnecessary cache misses and stale cache hits.

Fix that by adding .unique() and .sort() to the resolved dependency
list, and normalize the key format to "group:name:version". Also rename
getClasspath() to getResolvedDependencies() to better reflect its
purpose, and add a test verifying that transitive dependency changes
correctly invalidate the cache.

Fixes #327

chadlwilson reviewed (changes requested):
@jparise Thanks for trying to sort this out and digging into the root cause.

In your travels, did you perhaps figure out `CheckLicenseTask` doesn't have similar logic? I assume that is probably even more broken for the build cache than the report generation :-)

Although the problems didn't "start" with this change, it feels like this logic is increasingly coupled with the `ProjectReader` and `ConfigurationReader` internal logic for how it picks/filters relevant configurations and dependencies, which makes me a bit uncomfortable without refactoring this to ensure consistency and to extract/model the inputs better.

These pieces:

https://github.com/jk1/Gradle-License-Report/blob/eecb37ed1483da705ce56086c8787b182b443cdd/src/main/groovy/com/github/jk1/license/reader/ProjectReader.groovy#L47-L86

https://github.com/jk1/Gradle-License-Report/blob/e6a483d1d64e27e01da4bc7eef1c9b24ac1678e6/src/main/groovy/com/github/jk1/license/reader/ConfigurationReader.groovy#L44-L55

For example one gap seemingly predating this change, is that it doesn't seem to scope the cache key to the configurations actually **configured** for reporting, and it will resolve many more configurations than the "real" project. That was not such an issue before, I think, since it didn't actually resolve configurations - but I think now it will resolve (and download dependencies) even for configurations that are not configured which might have quite a big impact on projects like those for android with many `configuration`s and only some that are relevant.

Perhaps more importantly, I think this may also lead to more problems in multi-module projects, because it will cause Gradle to resolve configurations early (at configuration-time of any `ReportTask`s); which is often done from outside the given project's context, i.e making worse issues like https://github.com/jk1/Gradle-License-Report/issues/337

To move forward here, I am concerned that it might make things worse on Gradle 8/9 and break some of the workarounds listed at https://github.com/jk1/Gradle-License-Report/issues/337#issuecomment-3241442179 and https://github.com/jk1/Gradle-License-Report/issues/337#issuecomment-3649058336

The workarounds there tend to rely on late resolution and/or using `dependsOn` to ensure configurations are already resolved for "child" projects within the project context before `ReportTask` tries to work on them from an aggregate context.

To ensure we don't break things I think we probably need
- a test/more tests for reporting on some multi-module projects OR
- to resolve the configuration resolution issues in #337 before going further with this

WDYT @bdellegrazie ?

jparise commented:
I reworked this to a more targeted change that makes the cache key deterministic. It's less about transitive dependencies themselves now, given they're more reliably included by the other recent changes. The test I added for them is still valid though because we weren't covering that case before.

chadlwilson commented:
> I reworked this to a more targeted change that makes the cache key deterministic. It's less about transitive dependencies themselves now, given they're more reliably included by the other recent changes. The test I added for them is still valid though because we weren't covering that case before.

OK, I guess I need to look again and reset my brain as those other changes I had not looked at earlier and they seem somewhat invasive :-/
